### PR TITLE
Split messages

### PR DESCRIPTION
**Pull request by swalker2m**
This PR contains a couple of suggestions for discussion.

1) Split `StreamingMessages` by origin: `FromClient` and `FromServer`.  That allows matching on the appropriate subset, handling completely unexpected message types as an upstream parsing problem.

2) Manual codecs.  This may be controversial because it is admittedly _much more verbose_.  The idea with `StreamingMessage` though is that it defines a protocol.  If both sides use the same version of `clue` then it's fine that the codecs are automatically derived.  If that's not the case though then a seemingly innocent change to a class parameter name can break the other side if codecs are automatically derived.

### CODE REVIEW

File: model/shared/src/main/scala/clue/model/json/package.scala [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+  implicit val DecoderFromServer: Decoder[StreamingMessage.FromServer] =
+    List[Decoder[StreamingMessage.FromServer]](
+      decodeCaseObject("connection_ack", ConnectionAck),
+      Decoder[ConnectionError].widen,
+      decodeCaseObject("ka", ConnectionKeepAlive),
+      Decoder[Data].widen,
+      Decoder[Error].widen,
+      Decoder[Complete].widen
+    ).reduceLeft(_ or _)
```
Comment (rpiaggio) [inline model/shared/src/main/scala/clue/model/json/package.scala:187]: If decoding fails, will the `DecodingFailure` inform of all the decoders that were attempted (via `c.history` in `checkType`), or will it just report the last one?

File: project/Settings.scala [hunk 2]
```diff
@@ -5,14 +5,15 @@ import sbt.librarymanagement._
...
-    val circeGenericExtras = "0.13.0"
-    val log4Cats           = "1.1.1"
-    val scalaJSDom         = "1.1.0"
-    val sttpModel          = "1.1.4"
+    val cats                 = "2.1.1"
+    val catsEffect           = "2.1.4"
+    val catsTestkitScalaTest = "1.0.1"
+    val fs2                  = "2.4.2"
+    val circe                = "0.13.0"
+    val circeGenericExtras   = "0.13.0"
```
Comment (rpiaggio) [inline project/Settings.scala:13]: Do we still need `circeGenericExtras`?

**Comment by rpiaggio** [comment]
Thank you so much for all the improvements, including tests and comments.

About using manual codecs: Even though it's more verbose, it only needs to be done once and I agree that decoupling the implementation from the protocol is a good, future-proofing, thing.

BTW: is there a similar reason for manually defining the `Eq`s instead of deriving them? Otherwise, I would lean slightly towards derivation: if fields are added/renamed in case classes or members are added to sum types, there's less of a chance that things get broken if we forget to update the `Eq` instances accordingly.

**Comment by cquiroz** [comment]
I haven't looked at this on detail but manual codecs seem a lot of work. 
Considering that we can define exactly what fields we want we may have lots of small classes to create codecs for

Could we keep automatic or semi automatic derivation and use testing instead?
Have you looked at
https://github.com/circe/circe-golden

**Comment by rpiaggio** [comment]
`circe-golden` would ensure we don't break the protocol even if we use auto or semiauto derivation to take care of the boilerplate.

**Pull request merged**